core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -554,18 +554,22 @@ class KafkaApis(val requestChannel: RequestChannel,
 
         downConvertMagic.map { magic =>
           trace(s"Down converting records from partition $tp to message format version $magic for fetch request from $clientId")
-          val converted = data.records.downConvert(magic, fetchContext.getFetchOffset(tp).get, time)
-          updateRecordsProcessingStats(request, tp, converted.recordsProcessingStats)
-          new FetchResponse.PartitionData(data.error, data.highWatermark, FetchResponse.INVALID_LAST_STABLE_OFFSET,
-            data.logStartOffset, data.abortedTransactions, converted.records)
-        }
 
-      }.getOrElse(data)
+          // Because down-conversion is extremely memory intensive, we want to try and delay the down-conversion as much
+          // as possible. With KIP-283, we have the ability to lazily down-convert in a chunked manner. The lazy, chunked
+          // down-conversion always guarantees that at least one batch of messages is down-converted and sent out to the
+          // client.
+          val converted = new LazyDownConversionRecords(tp, data.records, magic, fetchContext.getFetchOffset(tp).get, Time.SYSTEM)
+          new FetchResponse.PartitionData[BaseRecords](data.error, data.highWatermark, FetchResponse.INVALID_LAST_STABLE_OFFSET,
+            data.logStartOffset, data.abortedTransactions, converted)
+        }
+      }.getOrElse(new FetchResponse.PartitionData[BaseRecords](data.error, data.highWatermark, data.lastStableOffset,
```

Comment:
You may have missed my previous comment, but this seems unnecessary. Can't we use `data` as we did in the original version?

Reply:
Sorry, I forgot to submit this comment.

I don't think we can use `data` directly because it is of type `FetchResponse.PartitionData<Records>` but we want `FetchResponse.PartitionData<BaseRecords>`. It might have been somewhat clearer if we had inheritance such that `FetchResponse.PartitionData<Records>` inherits from `FetchResponse.PartitionData<BaseRecords>`, in which case we could directly assign `<Records>` to `<BaseRecords>` instead of having to compose a new `<BaseRecords>` instance.